lib/wallet/wallet.js

```diff
@@ -2684,6 +2707,90 @@ class Wallet extends EventEmitter {
     return mtx;
   }
 
+  /**
+   * Make a renewal MTX for all expiring names.
+   * @param {MTX?} mtx
+   * @param {Number?} witnessSize
+   * @returns {MTX}
+   */
+
+  async makeRenewalAll(mtx, witnessSize) {
+    // Only allowed in makeBatch
+    assert(mtx, 'Batch MTX required for makeRenewalAll.');
+    assert(witnessSize, 'Witness size required for batch size estimation.');
+    const height = this.wdb.height + 1;
+    const network = this.network;
+    const names = await this.getNames();
+
+    let expiring = [];
+    for (const ns of names) {
+      // Easiest check is for expiring time, do that first
+      const {blocksUntilExpire} = ns.toStats(height, network);
+
+      if (!blocksUntilExpire || blocksUntilExpire < 0)
+        continue;
+      // TODO: Should this factor of 8 be user-configurable?
+      // About 90 days on main (1.75 years after REGISTER)
+      // 625 blocks on regtest (4375 blocks after REGISTER)
+      if (blocksUntilExpire >= network.names.renewalWindow / 8)
+        continue;
+
+      if (height < ns.renewal + network.names.treeInterval)
+        continue; // Can not renew yet
+
+      // Now do the db lookups to see if we own the name
+      const {hash, index} = ns.owner;
+      const coin = await this.getUnspentCoin(hash, index);
+      if (!coin)
+        continue;
+
+      if (!coin.covenant.isRegister()
+          && !coin.covenant.isUpdate()
+          && !coin.covenant.isRenew()
+          && !coin.covenant.isFinalize()) {
+        continue; // Name is not yet registered
+      }
+
+      expiring.push({ns, coin});
+    }
+
+    // Ignore in batches
+    if (!expiring.length)
+      return mtx;
+
+    // Sort by urgency, oldest/lowest renewal heights go first
+    expiring.sort((a, b) => {
+      return a.ns.renewal - b.ns.renewal;
+    });
+
+    // TODO: Should this factor of 6 be user-configurable?
+    // Enforce consensus limit per block at a maxmium
+    expiring = expiring.slice(0, consensus.MAX_BLOCK_RENEWALS / 6);
+
+    const renewalBlock = await this.wdb.getRenewalBlock();
+    for (const {ns, coin} of expiring) {
+      const output = new Output();
+      output.address = coin.address;
+      output.value = coin.value;
+      output.covenant.type = types.RENEW;
+      output.covenant.pushHash(ns.nameHash);
+      output.covenant.pushU32(ns.height);
+      output.covenant.pushHash(renewalBlock);
+
+      mtx.addOutpoint(new Outpoint(coin.hash, coin.index));
+      mtx.outputs.push(output);
+
+      // Keep batches below policy size limit
+      if (this.isOversizedBatch(mtx, witnessSize)) {
+        mtx.inputs.shift();
```

Comment:
totally! phew thanks

Reply:
kinda dont understand why this test didnt catch that: https://github.com/pinheadmz/hsd/blob/1a10677abd887c0a5d9622aae019c9d862d9f273/test/wallet-auction-test.js#L967-L969

Reply:
oh maybe because the renewal batch has a lower limit (consensus) so the shift/pop isnt even executed. it would be though, if the batch already had a bunch of other outputs. maybe ill write a bigger test with more mixed up actions